Change CI to create Debian Package to Release

AndrewQuijano opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [ ] I've documented or updated the documentation of every API function and struct this PR changes.
- [X] I've added tests that prove my fix is effective or that my feature works (if possible)

This is updating the CI, so every release, this would be tested.

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

A new feature is made such that every time a new release is created, a Debian package is created and uploaded to releases.

Approach:
Create a Debian folder, this contains everything needed to create a Debian package in the container. The container compiles and installs capstone, and creates the Debian package. The CI is updated to run the script which creates the container for upload release. 

Also, I did slight tweaks as your release CI needed to be updated too. I updated the release process to not use end of life uploading.
https://github.com/actions/upload-release-asset

Based on my hack in the Dockerfile, it seems to identify which files I need to copy in the Debian package
![image](https://github.com/user-attachments/assets/e9d5cbe7-c716-4d01-817e-0acf6e5d3f3e)

Note, I am copying (essentially should I copy /usr/lib/libcapstone.so too into the Debian Package?)
libcapstone.a
libcapstone.so
libcapstone.so.6

Proof the Debian package is created
![image](https://github.com/user-attachments/assets/30379f09-88e3-48af-b5ea-918d6125d87d)

...

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Or what test cases you added. Disassembly tests should be added in suite/cs_test/issues.cs -->

See thread below, a script is created to check that certain packages/libraries are in the Debian Package before adding to release.

...

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->
closes #2398 
...

AndrewQuijano committed:
Updating CI to create Debian package and version is assigned by tag
version. Also updating release CI to not use end-of-life workflows

Rot127 commented on .github/workflows/build_release.yml:
```diff
@@ -11,40 +11,19 @@ jobs:
       with:
         submodules: true
 
-    - name: archive
```
Why deleting the job to upload and create the archive?

AndrewQuijano replied:
Using set output is deprecated

https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

That is why I suspect tag 5.0.3 doesn't have the SHA and custom ZIP file. I'm wondering if this is necessary given the release process already tags the version.

https://github.com/capstone-engine/capstone/releases/tag/5.0.2
https://github.com/capstone-engine/capstone/releases/tag/5.0.3

AndrewQuijano replied:
If you are wondering, the changes should eventually look like this

https://github.com/panda-re/panda/releases/tag/v1.8.45

AndrewQuijano replied:
Plus there is the whole I am replacing the deprecated upload process with the newer version

AndrewQuijano replied:
Never mind ill try to fix this instead

Rot127 replied:
While reviewing and checking stuff, I think we should do this in another PR. This one gets too complex otherwise.
Also I think we should move to [CPack](https://cmake.org/cmake/help/book/mastering-cmake/chapter/Packaging%20With%20CPack.html). And then do a normal upload.
So even though this is deprecated since forever, please revert this change for now.

Rot127 commented on debian/Dockerfile:
```diff
@@ -0,0 +1,59 @@
+
+# Run the make.sh script and install
+WORKDIR /capstone/ 
+RUN ./make.sh
```
Please use the cmake build. make is considered deprecated. Check out `BUILDING.md`

AndrewQuijano replied:
Damn, seems like PANDA team was building it wrong the whole time, well I'll fix this in a second

https://github.com/panda-re/panda/blob/dev/Dockerfile#L38-L42

Rot127 replied:
All good. Makefiles still work. But we don't really take care of them anymore.

Rot127 commented on debian/control:
```diff
@@ -0,0 +1,8 @@
+Package: capstone
+Version: 3.1.0
```
~~This should be v6.0.0-alpha I think. Not sure how Debian likes it's versioning. But this is the latest released version.~~
Sorry, just saw you already do this. Nice! Please replace the version number with something like `<version-placeholder>` or similar. So we are not confused.

Rot127 commented on debian/setup.sh:
```diff
@@ -0,0 +1,35 @@
+    lsb_release -i -s 2>/dev/null
+}
+
+if [[ $(get_ubuntu_version) != "Ubuntu" ]]; then
```
Is this for Ubuntu only? Or also Debian?

AndrewQuijano replied:
technically should work for both, just copied this from Panda team though

Rot127 reviewed (changes requested):
Awesome! Please address the comments. We can test it by making a release in our own forks. And download and install the artifacts.

Rot127 commented:
> Note, I am copying (essentially should I copy /usr/lib/libcapstone.so too into the Debian Package?)

I think it should contain the static and shared lib file. You might need to build Capstone twice with cmake. Once for the static, once for creating the shared one.

AndrewQuijano commented:
Current diff looks like this, I suspect I can ignore the /sys stuff and I suspect I can ignore the stuff under /usr/local/lib/cmake/

```
7526a7527,7553
> /usr/local/include/capstone/loongarch.h
> /usr/local/include/capstone/riscv.h
> /usr/local/include/capstone/x86.h
> /usr/local/include/capstone/xcore.h
> /usr/local/include/capstone/tricore.h
> /usr/local/include/capstone/platform.h
> /usr/local/include/capstone/mips.h
> /usr/local/include/capstone/hppa.h
> /usr/local/include/capstone/bpf.h
> /usr/local/include/capstone/sh.h
> /usr/local/include/capstone/wasm.h
> /usr/local/include/capstone/systemz.h
> /usr/local/include/capstone/ppc.h
> /usr/local/include/capstone/aarch64.h
> /usr/local/include/capstone/tms320c64x.h
> /usr/local/include/capstone/arm.h
> /usr/local/include/capstone/mos65xx.h
> /usr/local/include/capstone/cs_operand.h
> /usr/local/include/capstone/m68k.h
> /usr/local/include/capstone/systemz_compatibility.h
> /usr/local/include/capstone/alpha.h
> /usr/local/include/capstone/evm.h
> /usr/local/include/capstone/m680x.h
> /usr/local/include/capstone/arm64.h
> /usr/local/include/capstone/xtensa.h
> /usr/local/include/capstone/capstone.h
> /usr/local/include/capstone/sparc.h
7549a7577,7582
> /usr/local/lib/pkgconfig/capstone.pc
> /usr/local/lib/cmake/capstone/capstone-targets.cmake
> /usr/local/lib/cmake/capstone/capstone-targets-release.cmake
> /usr/local/lib/cmake/capstone/capstone-config-version.cmake
> /usr/local/lib/cmake/capstone/capstone-config.cmake
> /usr/local/lib/libcapstone.so.6.0
10138a10172
> /usr/local/bin/cstool
21455,21523d21488
< /sys/devices/virtual/net/lo/uevent
< /sys/devices/virtual/net/lo/carrier_changes
< /sys/devices/virtual/net/lo/testing
< /sys/devices/virtual/net/lo/carrier
< /sys/devices/virtual/net/lo/dev_id
< /sys/devices/virtual/net/lo/carrier_down_count
< /sys/devices/virtual/net/lo/proto_down
< /sys/devices/virtual/net/lo/address
< /sys/devices/virtual/net/lo/operstate
< /sys/devices/virtual/net/lo/phys_switch_id
< /sys/devices/virtual/net/lo/link_mode
< /sys/devices/virtual/net/lo/dormant
< /sys/devices/virtual/net/lo/statistics/tx_errors
< /sys/devices/virtual/net/lo/statistics/rx_length_errors
< /sys/devices/virtual/net/lo/statistics/rx_packets
< /sys/devices/virtual/net/lo/statistics/tx_carrier_errors
< /sys/devices/virtual/net/lo/statistics/tx_dropped
< /sys/devices/virtual/net/lo/statistics/rx_missed_errors
< /sys/devices/virtual/net/lo/statistics/rx_over_errors
< /sys/devices/virtual/net/lo/statistics/tx_aborted_errors
< /sys/devices/virtual/net/lo/statistics/rx_crc_errors
< /sys/devices/virtual/net/lo/statistics/rx_frame_errors
< /sys/devices/virtual/net/lo/statistics/rx_nohandler
< /sys/devices/virtual/net/lo/statistics/tx_fifo_errors
< /sys/devices/virtual/net/lo/statistics/multicast
< /sys/devices/virtual/net/lo/statistics/tx_packets
< /sys/devices/virtual/net/lo/statistics/tx_window_errors
< /sys/devices/virtual/net/lo/statistics/rx_bytes
< /sys/devices/virtual/net/lo/statistics/collisions
< /sys/devices/virtual/net/lo/statistics/rx_dropped
< /sys/devices/virtual/net/lo/statistics/tx_bytes
< /sys/devices/virtual/net/lo/statistics/tx_heartbeat_errors
< /sys/devices/virtual/net/lo/statistics/rx_fifo_errors
< /sys/devices/virtual/net/lo/statistics/rx_errors
< /sys/devices/virtual/net/lo/statistics/tx_compressed
< /sys/devices/virtual/net/lo/statistics/rx_compressed
< /sys/devices/virtual/net/lo/mtu
< /sys/devices/virtual/net/lo/gro_flush_timeout
< /sys/devices/virtual/net/lo/carrier_up_count
< /sys/devices/virtual/net/lo/ifalias
< /sys/devices/virtual/net/lo/speed
< /sys/devices/virtual/net/lo/netdev_group
< /sys/devices/virtual/net/lo/napi_defer_hard_irqs
< /sys/devices/virtual/net/lo/phys_port_name
< /sys/devices/virtual/net/lo/ifindex
< /sys/devices/virtual/net/lo/broadcast
< /sys/devices/virtual/net/lo/type
< /sys/devices/virtual/net/lo/dev_port
< /sys/devices/virtual/net/lo/phys_port_id
< /sys/devices/virtual/net/lo/queues/tx-0/tx_maxrate
< /sys/devices/virtual/net/lo/queues/tx-0/xps_cpus
< /sys/devices/virtual/net/lo/queues/tx-0/byte_queue_limits/limit_min
< /sys/devices/virtual/net/lo/queues/tx-0/byte_queue_limits/limit
< /sys/devices/virtual/net/lo/queues/tx-0/byte_queue_limits/limit_max
< /sys/devices/virtual/net/lo/queues/tx-0/byte_queue_limits/hold_time
< /sys/devices/virtual/net/lo/queues/tx-0/byte_queue_limits/inflight
< /sys/devices/virtual/net/lo/queues/tx-0/tx_timeout
< /sys/devices/virtual/net/lo/queues/tx-0/xps_rxqs
< /sys/devices/virtual/net/lo/queues/tx-0/traffic_class
< /sys/devices/virtual/net/lo/queues/rx-0/rps_flow_cnt
< /sys/devices/virtual/net/lo/queues/rx-0/rps_cpus
< /sys/devices/virtual/net/lo/name_assign_type
< /sys/devices/virtual/net/lo/duplex
< /sys/devices/virtual/net/lo/addr_assign_type
< /sys/devices/virtual/net/lo/addr_len
< /sys/devices/virtual/net/lo/threaded
< /sys/devices/virtual/net/lo/tx_queue_len
< /sys/devices/virtual/net/lo/iflink
< /sys/devices/virtual/net/lo/flags
21592a21558,21626
> /sys/devices/virtual/net/lo/uevent
> /sys/devices/virtual/net/lo/carrier_changes
> /sys/devices/virtual/net/lo/testing
> /sys/devices/virtual/net/lo/carrier
> /sys/devices/virtual/net/lo/dev_id
> /sys/devices/virtual/net/lo/carrier_down_count
> /sys/devices/virtual/net/lo/proto_down
> /sys/devices/virtual/net/lo/address
> /sys/devices/virtual/net/lo/operstate
> /sys/devices/virtual/net/lo/phys_switch_id
> /sys/devices/virtual/net/lo/link_mode
> /sys/devices/virtual/net/lo/dormant
> /sys/devices/virtual/net/lo/statistics/tx_errors
> /sys/devices/virtual/net/lo/statistics/rx_length_errors
> /sys/devices/virtual/net/lo/statistics/rx_packets
> /sys/devices/virtual/net/lo/statistics/tx_carrier_errors
> /sys/devices/virtual/net/lo/statistics/tx_dropped
> /sys/devices/virtual/net/lo/statistics/rx_missed_errors
> /sys/devices/virtual/net/lo/statistics/rx_over_errors
> /sys/devices/virtual/net/lo/statistics/tx_aborted_errors
> /sys/devices/virtual/net/lo/statistics/rx_crc_errors
> /sys/devices/virtual/net/lo/statistics/rx_frame_errors
> /sys/devices/virtual/net/lo/statistics/rx_nohandler
> /sys/devices/virtual/net/lo/statistics/tx_fifo_errors
> /sys/devices/virtual/net/lo/statistics/multicast
> /sys/devices/virtual/net/lo/statistics/tx_packets
> /sys/devices/virtual/net/lo/statistics/tx_window_errors
> /sys/devices/virtual/net/lo/statistics/rx_bytes
> /sys/devices/virtual/net/lo/statistics/collisions
> /sys/devices/virtual/net/lo/statistics/rx_dropped
> /sys/devices/virtual/net/lo/statistics/tx_bytes
> /sys/devices/virtual/net/lo/statistics/tx_heartbeat_errors
> /sys/devices/virtual/net/lo/statistics/rx_fifo_errors
> /sys/devices/virtual/net/lo/statistics/rx_errors
> /sys/devices/virtual/net/lo/statistics/tx_compressed
> /sys/devices/virtual/net/lo/statistics/rx_compressed
> /sys/devices/virtual/net/lo/mtu
> /sys/devices/virtual/net/lo/gro_flush_timeout
> /sys/devices/virtual/net/lo/carrier_up_count
> /sys/devices/virtual/net/lo/ifalias
> /sys/devices/virtual/net/lo/speed
> /sys/devices/virtual/net/lo/netdev_group
> /sys/devices/virtual/net/lo/napi_defer_hard_irqs
> /sys/devices/virtual/net/lo/phys_port_name
> /sys/devices/virtual/net/lo/ifindex
> /sys/devices/virtual/net/lo/broadcast
> /sys/devices/virtual/net/lo/type
> /sys/devices/virtual/net/lo/dev_port
> /sys/devices/virtual/net/lo/phys_port_id
> /sys/devices/virtual/net/lo/queues/tx-0/tx_maxrate
> /sys/devices/virtual/net/lo/queues/tx-0/xps_cpus
> /sys/devices/virtual/net/lo/queues/tx-0/byte_queue_limits/limit_min
> /sys/devices/virtual/net/lo/queues/tx-0/byte_queue_limits/limit
> /sys/devices/virtual/net/lo/queues/tx-0/byte_queue_limits/limit_max
> /sys/devices/virtual/net/lo/queues/tx-0/byte_queue_limits/hold_time
> /sys/devices/virtual/net/lo/queues/tx-0/byte_queue_limits/inflight
> /sys/devices/virtual/net/lo/queues/tx-0/tx_timeout
> /sys/devices/virtual/net/lo/queues/tx-0/xps_rxqs
> /sys/devices/virtual/net/lo/queues/tx-0/traffic_class
> /sys/devices/virtual/net/lo/queues/rx-0/rps_flow_cnt
> /sys/devices/virtual/net/lo/queues/rx-0/rps_cpus
> /sys/devices/virtual/net/lo/name_assign_type
> /sys/devices/virtual/net/lo/duplex
> /sys/devices/virtual/net/lo/addr_assign_type
> /sys/devices/virtual/net/lo/addr_len
> /sys/devices/virtual/net/lo/threaded
> /sys/devices/virtual/net/lo/tx_queue_len
> /sys/devices/virtual/net/lo/iflink
> /sys/devices/virtual/net/lo/flags
29901c29935
< /before-install.txt
---
> /after-install.txt
30087a30122
> /capstone/build/install_manifest.txt
32467a32503
> /before-install.txt
```

AndrewQuijano commented:
Looks like it is working, going to push changes shortly

![image](https://github.com/user-attachments/assets/52b6a140-10ac-4191-ae30-efb73e9cd0f6)

AndrewQuijano commented:
Oh and @Rot127, assuming all this works as intended, could you make a capstone 5.0.3 Debian package or a 5.0.4 Debian package? The tool I use PANDA, for now, seems to only support version 5. The goal is to replace this with a one line to download the Debian package and install it.

Essentially, I do need a libcapstone.so.5 for the current capstone version
https://github.com/panda-re/panda/blob/dev/panda/scripts/install_ubuntu.sh#L109-L118

https://github.com/panda-re/panda/blob/dev/Dockerfile#L37-L41

AndrewQuijano commented:
Actually, I released I can test this on my own fork. Just be careful, you can only create tags say 5.0.4. If you have letters, think the debian part will break, just as a heads up. Works on 1.5.8 but not say v1.5.5

https://github.com/AndrewQuijano/capstone/releases/tag/1.5.8
https://github.com/AndrewQuijano/capstone/releases/tag/v1.5.5

Rot127 commented:
Thanks for all the work!

> Just be careful, you can only create tags say 5.0.4. If you have letters, think the debian part will break, just as a heads up. Works on 1.5.8 but not say v1.5.5

This is very good to know. Can you may lint the version string given to the script?
So for example:
- Remove letters from debian
- Or don't build for debian if other letters than a `v` prefix is given (e.g. in the case of `6.0.0a` for Alpha v6)?

We just need a way to have this checked, handled and documented in code. Because otherwise it will absolutely be forgotten at some point.

Rot127 commented:
> could you make a capstone 5.0.3 Debian package or a 5.0.4 Debian

Yes, we can release a new v5 version. Was on the list anyways. Just waiting for the TriCore bug to be resolved.

Rot127 commented on debian/Dockerfile:
```diff
@@ -0,0 +1,64 @@
+ARG VERSION="5.1"
```
Better not hard code a fall back version here. The script should always pass the version. So we have a single point where the version is specified.
Better check below if `VERSION` is given and fail if not.

AndrewQuijano replied:
I'll set to blank, this should break if you provide no argument

Rot127 commented on debian/Dockerfile:
```diff
@@ -0,0 +1,64 @@
+
+# Assume this is run from capstone/debian directory
+# Run in the root of the repo
+# docker build -f ./debian/Dockerfile -t packager .
```
Check this assumption in `setup.sh` please.

Rot127 commented on debian/setup.sh:
```diff
@@ -0,0 +1,36 @@
+# Copy deb file out of container to host
+docker run --rm -v $(pwd):/out packager bash -c "cp /capstone.deb /out"
+
+# Check which files existed before and after 'make install' was executed.
```
```suggestion
# Check which files existed before and after 'cmake install' was executed.
```

Rot127 commented on debian/control:
```diff
@@ -0,0 +1,8 @@
+Package: capstone
+Version: <version-placeholder>
+Architecture: all
+Maintainer: Wu ChenXu <capstone.engine@gmail.com>
```
Are you ok with adding yourself as maintainer here? You spent the largest time working on it for sure.

Rot127 replied:
If not, add me please.

AndrewQuijano replied:
I figure maintainer should be whoever works on Capstone, so what name should I put instead?

Rot127 replied:
Put `Rot127 <unisono@quyllur.org>` please. kabeor is pretty busy.

Rot127 commented on debian/Dockerfile:
```diff
@@ -0,0 +1,64 @@
+        fakeroot dpkg-dev dos2unix wget
+
+# Install CMake 3.15 or higher
+RUN wget https://github.com/Kitware/CMake/releases/download/v3.15.0/cmake-3.15.0-Linux-x86_64.sh && \
```
I think you can simply add cmake to the apt-get install above. No need for wget.

Rot127 replied:
This one is still unfixed though

AndrewQuijano replied:
I have tried downloading cmake via apt, apt only have cmake 3.12, this will break. I added a comment for this.

Rot127 commented on debian/Dockerfile:
```diff
@@ -0,0 +1,64 @@
+# Copy your project files into the container
+COPY . /capstone
+
+# Run the make.sh script and install
```
```suggestion
```

Rot127 commented on .github/workflows/build_release.yml:
```diff
@@ -11,9 +11,25 @@ jobs:
+      run: ./setup.sh ${{ github.event.release.tag_name }}
+
+    - name: Upload debian package to release
+      uses: softprops/action-gh-release@v2
```
I would like to not have more dependencies here. The deb package can be uploaded with:
```
      - uses: actions/upload-artifact@v4
        with:
          name: capstone.deb
          path:<path-t0-source>/capstone.deb
```
Or does `action-gh-release` also take care of publishing it in the debian repos?

AndrewQuijano replied:
This only uploads the debian package to release, this doesn't do anything with the debian apt repos.

Rot127 reviewed (changes requested):
Thanks a lot for the corrections. I spent way more time now on checking everything.

I ran it manually on a Debian VM and some stuff broke. Package building worked fine (as root: `./setup.sh 6.0.0`).

But there are a few problems:

The `capstone.pc` in the package is:

```sh
# After unpacking capstone.deb
> cat capstone/data/usr/local/lib/pkgconfig/capstone.pc

prefix=/usr/local
exec_prefix=${prefix}
libdir=${prefix}/lib
includedir=${prefix}/include

Name: capstone
Description: Capstone disassembly engine
Version: 6.0.
URL: https://www.capstone-engine.org/
archive=${libdir}/libcapstone.a
Libs: -L${libdir} -lcapstone
Libs.private: -L${libdir} -l:libcapstone.a
Cflags: -I${includedir} -I${includedir}/capstone
archs=
```

- The version number is not correct, it should end with a `0`. But ends with a `.`.
- `archs` is empty. I think it should be `arch` (no `s`) and be set to `x86_64`? But couldn't find docs about it. Just remove it, if you can't find an answer.
- `libcapstone.a` is not in the package. I think it is best, if you just fix `CMakeLists.txt` so it always compiles the static lib. Here the fix:
  ```diff
  diff --git a/CMakeLists.txt b/CMakeLists.txt

  index 345740c9..b8b01be6 100644
  --- a/CMakeLists.txt
  +++ b/CMakeLists.txt
  @@ -761,16 +761,23 @@ set(ALL_HEADERS
   set_property(GLOBAL PROPERTY VERSION ${PROJECT_VERSION})
   
   ## targets
  -add_library(capstone ${ALL_SOURCES} ${ALL_HEADERS})
  +add_library(capstone OBJECT ${ALL_SOURCES} ${ALL_HEADERS})
   add_library(capstone::capstone ALIAS capstone)
  +add_library(capstone_static STATIC $<TARGET_OBJECTS:capstone>)
  +# Use normal capstone name. Otherwise we get libcapstone_static.so
  +set_target_properties(capstone_static PROPERTIES OUTPUT_NAME "capstone")
   target_include_directories(capstone PUBLIC
       $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
   )
   set_property(TARGET capstone PROPERTY C_STANDARD 99)
   
   if(BUILD_SHARED_LIBS)
  -    target_compile_definitions(capstone PUBLIC CAPSTONE_SHARED)
  -    set_target_properties(capstone PROPERTIES
  +    set_property(TARGET capstone PROPERTY POSITION_INDEPENDENT_CODE 1)
  +    add_library(capstone_shared SHARED $<TARGET_OBJECTS:capstone>)
  +    # Use normal capstone name. Otherwise we get libcapstone_shared.so
  +    set_target_properties(capstone_shared PROPERTIES OUTPUT_NAME "capstone")
  +    target_compile_definitions(capstone_shared PUBLIC CAPSTONE_SHARED)
  +    set_target_properties(capstone_shared PROPERTIES
           VERSION ${PROJECT_VERSION}
           SOVERSION ${PROJECT_VERSION_MAJOR}
       )
  ```
- `cstool` doesn't work after installing the package, because `ldconfig` is not run (and `/usr/local/lib` is not in the default lib search path). [This](https://askubuntu.com/questions/49814/io-postinst-must-call-ldconfig-when-creating-a-package) seems to be the way to run `ldconfig` after install.
- Please add a test script which installs the package and checks everything is correct (version number, libs are there, libs report the same version, cstool works etc). It would be good to have this in a separated job.

Sorry for being super nit-picky with this one. We had quite some problems with the Python release packages. And it is hard to fix, because we need to do a fix-release after changing it. Very stressful experience. This is why I ask you for so much stuff.

AndrewQuijano commented:
Hello, 
I just got your changes in, seems like libcapstone.a is no longer appearing, not sure how to fix this

```
> /usr/local/include/capstone/loongarch.h
> /usr/local/include/capstone/riscv.h
> /usr/local/include/capstone/x86.h
> /usr/local/include/capstone/xcore.h
> /usr/local/include/capstone/tricore.h
> /usr/local/include/capstone/platform.h
> /usr/local/include/capstone/mips.h
> /usr/local/include/capstone/hppa.h
> /usr/local/include/capstone/bpf.h
> /usr/local/include/capstone/sh.h
> /usr/local/include/capstone/wasm.h
> /usr/local/include/capstone/systemz.h
> /usr/local/include/capstone/ppc.h
> /usr/local/include/capstone/aarch64.h
> /usr/local/include/capstone/tms320c64x.h
> /usr/local/include/capstone/arm.h
> /usr/local/include/capstone/mos65xx.h
> /usr/local/include/capstone/cs_operand.h
> /usr/local/include/capstone/m68k.h
> /usr/local/include/capstone/systemz_compatibility.h
> /usr/local/include/capstone/alpha.h
> /usr/local/include/capstone/evm.h
> /usr/local/include/capstone/m680x.h
> /usr/local/include/capstone/arm64.h
> /usr/local/include/capstone/xtensa.h
> /usr/local/include/capstone/capstone.h
> /usr/local/include/capstone/sparc.h
7549a7577,7581
> /usr/local/lib/pkgconfig/capstone.pc
> /usr/local/lib/cmake/capstone/capstone-targets.cmake
> /usr/local/lib/cmake/capstone/capstone-targets-release.cmake
> /usr/local/lib/cmake/capstone/capstone-config-version.cmake
> /usr/local/lib/cmake/capstone/capstone-config.cmake
10138a10171
> /usr/local/bin/cstool
```

Rot127 commented:
Oups, forgot to update the install targets:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 345740c9..c8c10880 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -878,7 +885,13 @@ if(CAPSTONE_INSTALL)
         DESTINATION ${CAPSTONE_CMAKE_CONFIG_INSTALL_DIR}
     )
 
-    install(TARGETS capstone
+    if(BUILD_SHARED_LIBS)
+        set(LIB_INSTALL_TARGETS capstone_static capstone_shared)
+    else()
+        set(LIB_INSTALL_TARGETS capstone_static)
+    endif()
+
+    install(TARGETS ${LIB_INSTALL_TARGETS}
             EXPORT capstone-targets
             RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
             LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
```

AndrewQuijano commented:
OK, I just need to fix capstone.pc version and removing arch and I am done, but in the meantime please test what I got so far

Rot127 commented on debian/check_capstone.sh:
```diff
@@ -0,0 +1,67 @@
+  echo "libcapstone.a not found in the package!"
+  rm -rf "$TEMP_DIR"
+  exit 1
+fi
```
Also check for `.so` file please.

Rot127 commented on debian/check_capstone.sh:
```diff
@@ -0,0 +1,67 @@
+  exit 1
+fi
+
+# Check if the archs field is present in the capstone.pc file
+if grep -q "^archs=" "$CAPSTONE_PC"; then
+  echo "archs field found in capstone.pc file!"
+  rm -rf "$TEMP_DIR"
+  exit 1
+fi
+
```
```suggestion
```

Sorry, I meant you can remove `archs=@CAPSTONE_ARCHITECTURES@` from `capstone.pc.in`. Because it isn't used.

AndrewQuijano commented:
Hm, this is tricky

![image](https://github.com/user-attachments/assets/2affbe79-65b1-4904-83ad-33521ca50364)

Rot127 commented:
The workflow fails in my test release:
https://github.com/Rot127/capstone/actions/runs/11461606430/job/31890993552

AndrewQuijano commented:
Ah it only works with one number at the moment, e. g. 5.7.1, I think it breaks with like 55.2.4

Rot127 commented:
Numbers seem fine. I think it is because it passes `./debian/capstone.deb` to the check script. But the working dir is already `./debian/`

AndrewQuijano commented:
yeah just patched it

AndrewQuijano commented:
Oh I forgot to pull out debian package because I was debugging the library issue. This should work end to end now

Rot127 commented on debian/Dockerfile:
```diff
@@ -0,0 +1,69 @@
+# Assume this is run from capstone/debian directory
+# Run in the root of the repo
+# docker build -f ./debian/Dockerfile -t packager .
+FROM debian:buster-slim
```
```suggestion
FROM debian:bookworm-slim
```

Just saw it. buster is EOL since 2 years.

Rot127 commented:
I need to go. Will take a look at it tomorrow again. Thanks for fixing it so quickly.

AndrewQuijano commented:
No worries! Thanks for your fast responses! Really look forward to updating PANDA with the package. Also QQ where did you find buster-slim is EOL? Figure I might as well update PANDA's Debian package thing too, but figure I want to to document it in the PR

AndrewQuijano commented:
Looks like this works!

https://github.com/AndrewQuijano/capstone/actions/runs/11462174350/job/31892948396

https://github.com/AndrewQuijano/capstone/releases/tag/v1.5.9

Rot127 commented:
> where did you find buster-slim is EOL?

Debian Buster is EOL since two years. Hence we don't have to build on it. This is also likely the reason you could not install cmake via apt.

Rot127 committed:
Clear up usage of static libraries.

- Python bindings only use the dynamic lib. But built and copied the static ones sometimes nonetheless.
- Add toggles to build only static, static/dyn or only dynamic.

Rot127 commented:
@AndrewQuijano Check https://github.com/AndrewQuijano/capstone/pull/1 please

AndrewQuijano commented:
Clear up usage of static libraries.
- Python bindings only use the dynamic lib. But built and copied the static ones sometimes nonetheless.
- Add toggles to build only static, static/dyn or only dynamic.

Install cmake via apt.
cmake >3.12 was not present in Debian Buster.

AndrewQuijano commented:
@Rot127 just accepted your PR

AndrewQuijano commented:
Thanks for your help with this! Just curious, when would you expect v 5.0.4 of capstone Debian Package be ready by? Figure I can start working on the PR for PANDA to use the debian package. I'm wondering if there might come a day you might have an apt repository? 

https://earthly.dev/blog/creating-and-hosting-your-own-deb-packages-and-apt-repo/

Rot127 commented:
Nice! Lgtm. I fix the failing test tomorrow.
@kabeor Please review this one.

Rot127 commented:
> I'm wondering if there might come a day you might have an apt repository?

It is not planned. We need maintainers for this. And we are already too busy with all the other stuff to do it on our own.

> when would you expect v 5.0.4 of capstone Debian Package be ready by? 

https://github.com/capstone-engine/capstone/pull/2523 is the PR which has to be done. I review it tomorrow as well. Then we need to merge it into v5.
Depending how much we have to change I guess 1-6 weeks. billow and me are pretty busy these days. So hard to say.

AndrewQuijano commented:
btw I noticed it still says request changes blocking the merge:
https://github.com/capstone-engine/capstone/pull/2521#pullrequestreview-2384823865

1. Do you still want me to remove the 'archs' from the pkconfig?
2. Safe to say the capstone debian package worked without needing ldconfig?

Rot127 commented on debian/check_capstone.sh:
```diff
@@ -0,0 +1,68 @@
+fi
+
+# Check if libcapstone.so is included in the package
+LIBCAPSTONE_SO="$TEMP_DIR/usr/local/lib/libcapstone.a"
```
```suggestion
LIBCAPSTONE_SO="$TEMP_DIR/usr/local/lib/libcapstone.so"
```

Rot127 commented on debian/postinst:
```diff
@@ -0,0 +1,7 @@
+# Update the shared library cache
+ldconfig
+
+exit 0
```
```suggestion
```

Otherwise we never catch ldconfig errors.

Rot127 reviewed (changes requested):
Please just address the last two small things.


    Do you still want me to remove the 'archs' from the pkconfig?

Actually yeah. It is not used outside of the Make build. It was added to have a list of supported architectures. But this information doesn't belong there. And I couldn't find documentation about this `archs` field. So it is likely not standard anyways.

    Safe to say the capstone debian package worked without needing ldconfig?

At least in my fresh Debian VM it didn't work otherwise. Feel free to test it again.

AndrewQuijano commented:
@Rot127 Just made your suggested changes, let me know about if the ldconfig works/the package installation works as intended. Look forward to merging the PR!

AndrewQuijano commented:
@Rot127, let me know when this PR is to be merged.

I made one last change, and provided more context in https://github.com/capstone-engine/capstone/issues/2398. I changed the package name from 'capstone.deb' to 'libcapstone-dev.deb' so it matches the name of the apt-package. Just to confirm, this debian package essentially would be both libcapstone5.deb and libcapstone-dev.deb package from apt? I suspect the answer is yes given that both library file is added, and the /usr/local/include is populated with headers.

And please confirm the package works as expected. Essentially, PANDA would have required both the capstone library, and does rely on capstone for development.
https://www.capstone-engine.org/documentation.html

Rot127 reviewed (approved):
I am sorry, I thought I already approved this one. But I can't merge anyways. The review of @kabeor is required. I already reminded him a few days ago.

> libcapstone5.deb and libcapstone-dev.deb package from apt?

Yeah, you are right. We should have a `libcapstone6` (not `libcapstone5`) and `libcapstone-dev` package. But let's keep the PR like this for now please. Because it is already large enough.
Also, as long as we have not submitted it to the Debian/Ubuntu repos, we don't need the distinction yet.

If you want to take care of it, feel free to open a new PR of course.

> And please confirm the package works as expected.

Works for me! Tested it with a Whonix VM. See https://github.com/Rot127/capstone/releases/tag/v99.99.103

kabeor commented on CMakeLists.txt:
```diff
@@ -55,7 +55,8 @@ endif()
 # to configure the options specify them in the command line or change them in the cmake UI.
 # Don't edit the makefile!
 option(BUILD_SHARED_LIBS "Build shared library" OFF)
```
I guess change this to CAPSTONE_BUILD_SHARED_LIBS is better, how do you think @Rot127 ?

kabeor commented on CMakeLists.txt:
```diff
@@ -55,7 +55,8 @@ endif()
 option(BUILD_SHARED_LIBS "Build shared library" OFF)
-option(CAPSTONE_BUILD_STATIC_RUNTIME "Embed static runtime" ${BUILD_SHARED_LIBS})
+option(BUILD_STATIC_LIBS "Build static library" ON)
+option(BUILD_STATIC_RUNTIME "Embed static MSVC runtime (Windows only). Always set if BUILD_SHARED_LIBS=ON" ${BUILD_SHARED_LIBS})
```
Maybe CAPSTONE_BUILD_STATIC_MSVC_RUNTIME?

Rot127 replied:
Yes to both. But let's update it in another PR.

Rot127 replied:
Because it is not related to the Debian package. I'll open one once this is merged.

kabeor reviewed (changes requested):
Everything else are LGTM.

kabeor reviewed (approved):
Thank you, this is amazing.
